# Diagnosing a model that isn't working

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/byu-matrix-lab/torchlingo/blob/main/docs/docs/tutorials/06-diagnosing-failures.ipynb)

Every other tutorial shows you something that works. This one breaks things on
purpose, so that when your own model misbehaves you have a *procedure* instead
of a hunch.

The companion page [When it fails](../concepts/when-it-fails.md) lists real
failures and what caused each one. Reading that page tells you the answers. This
notebook makes you watch the checks fire.

By the end you will have:

1. run five checks against a healthy model, as a control;
2. seen each one fire on a break made on purpose;
3. the procedure table at the end, to look your own symptom up in.

It runs on a CPU in a few minutes and downloads nothing. In Colab, run the first
two code cells first: they install TorchLingo and check the setup.

<!-- GENERATED:purpose - change the torchlingo metadata, not this cell -->

**TorchLingo tutorial** — assigned reading for Lecture 9 (*Handling Morphology and Terminology in NMT*). A head start on **A8**.

In [1]:
# Install TorchLingo in Colab. Nothing to uncomment. Pip's "you may need to restart the
# kernel" does not apply to a first install; if the install fails, the import below stops here.
import sys
if "google.colab" in sys.modules:
    %pip install --quiet "torchlingo>=0.2.2"
import torchlingo

In [2]:
# Check the setup: the TorchLingo and PyTorch versions and the device, printed here so that
# anything wrong shows up now rather than ten cells from now.
from torchlingo.colab import setup

env = setup()

TorchLingo 0.2.3 | PyTorch 2.13.0 | device cpu | not in Colab


## The order matters

Work top to bottom. Each question is cheaper to answer than the one below it,
and a *yes* at any level makes everything below it meaningless. There is no
point tuning a learning rate on a corpus whose two sides do not match.

| # | Question | Cost to check |
| - | -------- | ------------- |
| 1 | Is the data what you think it is? | seconds |
| 2 | Is the model learning *anything*? | one backward pass |
| 3 | Is it learning the *wrong* thing? | one training run |
| 4 | Is the measurement lying? | seconds |
| 5 | Is it the environment? | seconds |

**Use this as a lookup.** When your own run misbehaves, find your symptom in the
table at the end, run that one question's cells (after the setup and the
control), and read its Recognize and Fix. Reading top to bottom once is worth
doing; rereading it is not.

Every check comes from [`torchlingo.diagnostics`](../reference/diagnostics.md),
so you can `import` the same functions into your own work. Each returns a
`CheckResult` that prints as **PASS** or **FAIL** against a stated threshold, and
is truthy when it passed, so `assert check_eval_mode(model)` reads as a guard.

## Setup: a toy corpus we can break

Real translation needs hours of training, so you cannot tell a bug from an
undertrained model. Instead, a synthetic English/Spanish task that a small
Transformer solves in about a minute: when it *doesn't*, we broke something.

Sentences look like `Maria sees 42 old cats` / `Maria ve 42 viejos gatos`: word
for word, so a correct corpus has a length correlation of exactly 1.0, and names
and numbers pass through untranslated, giving the alignment checks anchors to
look for.

In [3]:
# Setup: run this cell, no need to read it. It builds the toy corpus described above.
import inspect
import math
import random
import warnings
from functools import partial

import pandas as pd
import torch
from torch import nn, optim
from torch.utils.data import DataLoader, Dataset

from torchlingo.config import Config
from torchlingo.data_processing.batching import collate_fn
from torchlingo.diagnostics import (
    CheckResult,
    check_contamination,
    check_eval_mode,
    check_generalization,
    check_gradients,
    check_loss_moved,
    gradient_report,
    uniform_loss,
)
from torchlingo.evaluation import compute_bleu
from torchlingo.inference import greedy_decode
from torchlingo.models import SimpleTransformer
from torchlingo.preprocessing.alignment import diagnose_alignment, shuffle_target_side
from torchlingo.training import train_model

warnings.filterwarnings("ignore")

NAMES = ["Maria", "Juan", "Ana", "Pedro", "Elena", "Carmen", "Diego", "Rosa"]
VERB = {"sees": "ve", "wants": "quiere", "finds": "encuentra", "likes": "aprecia"}
NOUN = {"cats": "gatos", "dogs": "perros", "birds": "pajaros", "fish": "peces"}
ADJ = {"old": "viejos", "young": "jovenes", "big": "grandes", "small": "pequenos"}


def make_corpus(n: int, seed: int) -> pd.DataFrame:
    """Build n parallel sentence pairs. Length varies so correlation is measurable."""
    rng = random.Random(seed)
    rows = []
    for _ in range(n):
        name = rng.choice(NAMES)
        verb = rng.choice(list(VERB))
        number = rng.randint(10, 99)
        adjs = rng.sample(list(ADJ), rng.randint(0, 2))
        noun = rng.choice(list(NOUN))
        source = [name, verb, str(number), *adjs, noun]
        target = [name, VERB[verb], str(number), *[ADJ[a] for a in adjs], NOUN[noun]]
        rows.append((" ".join(source), " ".join(target)))
    return pd.DataFrame(rows, columns=["src", "tgt"])


train_df = make_corpus(3000, seed=0)
val_df = make_corpus(400, seed=99)
train_df.head(3)

,src,tgt
0,Diego likes 15 small fish,Diego aprecia 15 pequenos peces
1,Elena likes 55 young big dogs,Elena aprecia 55 jovenes grandes perros
2,Elena wants 22 big small dogs,Elena quiere 22 grandes pequenos perros


In [4]:
# Setup: run this cell, no need to read it. A vocabulary, a dataset and a small model for
# the toy corpus; new_model() and translate() are what the questions below call.
class Vocab:
    """A word-level vocabulary over both languages at once."""

    def __init__(self, frames):
        tokens = set()
        for frame in frames:
            for column in ("src", "tgt"):
                for sentence in frame[column]:
                    tokens.update(sentence.split())
        self.itos = ["<pad>", "<sos>", "<eos>"] + sorted(tokens)
        self.stoi = {t: i for i, t in enumerate(self.itos)}
        self.pad_idx, self.sos_idx, self.eos_idx = 0, 1, 2

    def __len__(self):
        return len(self.itos)

    def encode(self, text):
        return [self.sos_idx] + [self.stoi[w] for w in text.split()] + [self.eos_idx]

    def decode(self, ids):
        return " ".join(self.itos[i] for i in ids if i > self.eos_idx)


class Pairs(Dataset):
    def __init__(self, frame, vocab):
        self.frame = frame.reset_index(drop=True)
        self.vocab = vocab

    def __len__(self):
        return len(self.frame)

    def __getitem__(self, i):
        row = self.frame.iloc[i]
        return (
            torch.tensor(self.vocab.encode(row.src)),
            torch.tensor(self.vocab.encode(row.tgt)),
        )


vocab = Vocab([train_df, val_df])

# label_smoothing is off so that 0.0 means "perfect" and ln(V) means "guessing".
# With the default 0.1 a converged model still sits near 0.6 on this vocabulary,
# which makes both reference points harder to reason about.
CFG = Config(batch_size=64, label_smoothing=0.0, pad_idx=0, sos_idx=1, eos_idx=2)
UNIFORM_LOSS = uniform_loss(len(vocab))


def loader(frame, batch_size=64, shuffle=True):
    return DataLoader(
        Pairs(frame, vocab),
        batch_size=batch_size,
        shuffle=shuffle,
        collate_fn=partial(collate_fn, pad_idx=vocab.pad_idx),
    )


def new_model(seed=0):
    torch.manual_seed(seed)
    return SimpleTransformer(
        src_vocab_size=len(vocab),
        tgt_vocab_size=len(vocab),
        d_model=128,
        n_heads=4,
        num_encoder_layers=2,
        num_decoder_layers=2,
        d_ff=256,
        dropout=0.1,
    )


def translate(model, sentences, max_len=12):
    model.eval()
    ids = [vocab.encode(s) for s in sentences]
    width = max(len(x) for x in ids)
    src = torch.tensor([x + [vocab.pad_idx] * (width - len(x)) for x in ids])
    return [vocab.decode(o) for o in greedy_decode(model, src, max_len=max_len, config=CFG)]


print(f"vocabulary: {len(vocab)} types")
print(f"a model that has learned nothing scores ln(V) = {UNIFORM_LOSS:.3f}")

vocabulary: 125 types
a model that has learned nothing scores ln(V) = 4.828


## The control: a model that works

Before breaking anything we need to know what healthy looks like on this task.
Keep these two numbers — every failure below is measured against them.

This takes about a minute.

In [5]:
healthy = new_model()
healthy_run = train_model(
    healthy,
    loader(train_df),
    loader(val_df, shuffle=False),
    num_epochs=20,
    config=CFG,
    optimizer=optim.Adam(healthy.parameters(), lr=1e-3),
)

HEALTHY_TRAIN = healthy_run.train_losses[-1]
HEALTHY_VAL = healthy_run.val_losses[-1]
print(f"\ntrain {HEALTHY_TRAIN:.3f}   val {HEALTHY_VAL:.3f}   (guessing would be {UNIFORM_LOSS:.3f})")
for source, output in zip(val_df.src[:3], translate(healthy, list(val_df.src[:3]))):
    print(f"  {source:32} -> {output}")

Epoch 1/20 | Train: 5.2225 | Val: 4.9287


Epoch 2/20 | Train: 4.6007 | Val: 4.0425


Epoch 3/20 | Train: 3.9316 | Val: 3.4984


Epoch 4/20 | Train: 3.5191 | Val: 3.1023


Epoch 5/20 | Train: 3.1546 | Val: 2.6899


Epoch 6/20 | Train: 2.7770 | Val: 2.2938


Epoch 7/20 | Train: 2.4024 | Val: 1.9241


Epoch 8/20 | Train: 2.0529 | Val: 1.5647


Epoch 9/20 | Train: 1.6933 | Val: 1.2546


Epoch 10/20 | Train: 1.3780 | Val: 1.0637


Epoch 11/20 | Train: 1.1789 | Val: 0.9670


Epoch 12/20 | Train: 1.0651 | Val: 0.9027


Epoch 13/20 | Train: 0.9741 | Val: 0.8491


Epoch 14/20 | Train: 0.9119 | Val: 0.7981


Epoch 15/20 | Train: 0.8530 | Val: 0.7502


Epoch 16/20 | Train: 0.7924 | Val: 0.6468


Epoch 17/20 | Train: 0.6845 | Val: 0.4576


Epoch 18/20 | Train: 0.5233 | Val: 0.2765


Epoch 19/20 | Train: 0.3715 | Val: 0.1592


Epoch 20/20 | Train: 0.2653 | Val: 0.0987

train 0.265   val 0.099   (guessing would be 4.828)
  Diego likes 35 young old dogs    -> Diego aprecia 35 viejos jovenes perros
  Ana sees 42 small big cats       -> Ana ve 42 grandes pequenos gatos
  Rosa wants 63 young small fish   -> Rosa quiere 63 pequenos jovenes peces


## Question 1 — Is the data what you think it is?

This is first because it is the cheapest to check and the most expensive to miss.
A corpus whose two sides drifted out of alignment still loads, still batches,
still trains, and still produces a falling loss curve. Nothing raises.

`shuffle_target_side` exists to reconstruct exactly that failure on any corpus,
so you can watch the checks catch it.

**The diagnostic** is `diagnose_alignment`, which runs two independent checks:

- **Length correlation.** Translations preserve length approximately. Genuinely
  parallel text correlates near 0.97; unrelated pairs near 0.0.
- **Anchor agreement.** Names and numbers survive translation. The fraction of
  rows whose two sides share at least one should be high.

`looks_aligned()` requires correlation >= 0.80 **and** agreement >= 0.25.

In [6]:
def check_alignment(frame, label):
    """Question 1 has no entry in `torchlingo.diagnostics`, deliberately.

    It needs no model, so it lives with the corpus tools in
    `torchlingo.preprocessing.alignment`. Wrapping it in a `CheckResult` here
    only makes all five questions print the same way.
    """
    result = diagnose_alignment(frame)
    return CheckResult(
        f"corpus alignment ({label})",
        result.looks_aligned(),
        f"length_correlation={result.length_correlation:.3f} (need >=0.80)   "
        f"anchor_agreement={result.anchor_agreement:.3f} (need >=0.25)",
    )


broken_df = shuffle_target_side(train_df)
print(check_alignment(train_df, "as built"))
print(check_alignment(broken_df, "after shuffle_target_side"))

print("\nthe same row, both ways:")
print(f"  intact : {train_df.src[0]!r} -> {train_df.tgt[0]!r}")
print(f"  broken : {broken_df.src[0]!r} -> {broken_df.tgt[0]!r}")

[PASS] corpus alignment (as built) — length_correlation=1.000 (need >=0.80)   anchor_agreement=1.000 (need >=0.25)
[FAIL] corpus alignment (after shuffle_target_side) — length_correlation=0.013 (need >=0.80)   anchor_agreement=0.118 (need >=0.25)

the same row, both ways:
  intact : 'Diego likes 15 small fish' -> 'Diego aprecia 15 pequenos peces'
  broken : 'Diego likes 15 small fish' -> 'Elena aprecia 55 jovenes grandes perros'


### What it costs to skip this check

Train on the broken corpus and nothing announces a problem — the loss falls and
the curve looks like training.

Give it the same 20 epochs the healthy model got, changing exactly one variable
(the corpus), and the difference is unmistakable.

In [7]:
on_broken = new_model()
broken_run = train_model(
    on_broken,
    loader(broken_df),
    loader(val_df, shuffle=False),
    num_epochs=20,
    config=CFG,
    optimizer=optim.Adam(on_broken.parameters(), lr=1e-3),
)

print(f"\nlast 5 epochs on the broken corpus: {[round(x, 3) for x in broken_run.train_losses[-5:]]}")
print(f"  it has flattened out at  train {broken_run.train_losses[-1]:.3f}")
print(f"  healthy, same 20 epochs: train {HEALTHY_TRAIN:.3f}")
print("\nwhat it produces:")
for source, output in zip(val_df.src[:3], translate(on_broken, list(val_df.src[:3]))):
    print(f"  {source:32} -> {output}")

Epoch 1/20 | Train: 5.2421 | Val: 4.9416


Epoch 2/20 | Train: 4.6296 | Val: 4.1051


Epoch 3/20 | Train: 4.0075 | Val: 3.6322


Epoch 4/20 | Train: 3.6485 | Val: 3.3412


Epoch 5/20 | Train: 3.3848 | Val: 3.0693


Epoch 6/20 | Train: 3.1200 | Val: 2.8274


Epoch 7/20 | Train: 2.8831 | Val: 2.6184


Epoch 8/20 | Train: 2.6724 | Val: 2.4308


Epoch 9/20 | Train: 2.4620 | Val: 2.2534


Epoch 10/20 | Train: 2.2992 | Val: 2.1569


Epoch 11/20 | Train: 2.2036 | Val: 2.1057


Epoch 12/20 | Train: 2.1569 | Val: 2.0775


Epoch 13/20 | Train: 2.1185 | Val: 2.0613


Epoch 14/20 | Train: 2.0966 | Val: 2.0518


Epoch 15/20 | Train: 2.0799 | Val: 2.0408


Epoch 16/20 | Train: 2.0622 | Val: 2.0366


Epoch 17/20 | Train: 2.0504 | Val: 2.0282


Epoch 18/20 | Train: 2.0477 | Val: 2.0246


Epoch 19/20 | Train: 2.0365 | Val: 2.0212


Epoch 20/20 | Train: 2.0259 | Val: 2.0204

last 5 epochs on the broken corpus: [2.062, 2.05, 2.048, 2.036, 2.026]
  it has flattened out at  train 2.026
  healthy, same 20 epochs: train 0.265

what it produces:
  Diego likes 35 young old dogs    -> Carmen aprecia 11 grandes jovenes perros
  Ana sees 42 small big cats       -> Diego aprecia 11 grandes perros jovenes
  Rosa wants 63 young small fish   -> Carmen aprecia 11 grandes perros jovenes


**Recognize it:** the loss falls, then flattens out far above where a comparable
run reaches — and the outputs are built from target-language vocabulary that has
little to do with the input. Names and numbers, which a working model copies
straight through, come out wrong or missing entirely.

That is the signature of a model that learned the target language and could not
learn the mapping, because in this corpus there isn't one to learn.

**Fix it:** run `diagnose_alignment` on every corpus *before* training, and treat
`looks_aligned() == False` as a hard stop. Anything landing between the
thresholds is worth reading by hand. Realigning a corpus that has drifted is its
own job — see `gale_church_align` in `torchlingo.preprocessing.alignment`.

## Question 2 — Is the model learning *anything*?

There is a specific number that means "learned nothing": the loss of a model that
guesses uniformly over the vocabulary, `ln(V)`. A loss sitting there, flat, is
not a slow model — it is a disconnected one.

Here we break it the most common way there is: a learning rate of zero.

The flat number below sits a little *above* `ln(V)`, because a freshly built model
is not quite uniform: it is confidently wrong about some tokens. What identifies
the failure is that the number does not *move*.

In [8]:
stuck = new_model()
stuck_run = train_model(
    stuck,
    loader(train_df),
    loader(val_df, shuffle=False),
    num_epochs=3,
    config=CFG,
    optimizer=optim.Adam(stuck.parameters(), lr=0.0),  # <-- the bug
)

# check_loss_moved takes the loss history rather than a TrainResult, so it also
# works on a hand-rolled loop or a slice of epochs. Pass vocab_size and it
# quotes ln(V) alongside, which is what "learned nothing" scores.
print()
print(check_loss_moved(stuck_run.train_losses, vocab_size=len(vocab)))
print(check_loss_moved(healthy_run.train_losses, vocab_size=len(vocab)))

Epoch 1/3 | Train: 5.3408 | Val: 5.3712


Epoch 2/3 | Train: 5.3411 | Val: 5.3712


Epoch 3/3 | Train: 5.3365 | Val: 5.3712

[FAIL] loss moved — first=5.341 last=5.337 drop=0.0043 (need >=0.05); guessing scores 4.828
[PASS] loss moved — first=5.223 last=0.265 drop=4.9572 (need >=0.05); guessing scores 4.828


### Localizing it: which parameters can actually learn?

"Loss is flat" tells you something is wrong, not *what*. One backward pass
answers that, and it sorts every parameter into three buckets:

| Bucket | Meaning | Usual cause |
| ------ | ------- | ----------- |
| **frozen** | `requires_grad=False` | someone froze a submodule and forgot |
| **dead** | gradient is `None` or all zeros | the graph was detached, or the output is unused |
| **live** | gradient is nonzero | healthy |

It needs no training, only a freshly built model. Below we freeze the encoder and
watch it get named.

In [9]:
# This is the one check with no cheaper substitute, so read it rather than just
# calling it. It is the library's own source, printed from the installed module:
print(inspect.getsource(gradient_report))

def gradient_report(model: nn.Module, loss: torch.Tensor) -> GradientReport:
    """Sort every parameter into frozen, dead, or live by one backward pass.

    Call this when :func:`check_loss_moved` fails. It answers the question that
    a flat loss curve cannot: *which* parameters were never going to move.

    Existing gradients are cleared before the backward pass, so a stale
    ``.grad`` left over from earlier training cannot make a dead parameter look
    live. This consumes the graph, so pass a loss you are not going to call
    ``backward()`` on yourself.

    Args:
        model (nn.Module): The model whose parameters to inspect.
        loss (torch.Tensor): A scalar loss computed from ``model``, still
            attached to its graph.

    Returns:
        GradientReport: The three buckets, by parameter name.

    Raises:
        ValueError: If ``loss`` is not a scalar, or has no grad history — both
            of which are themselves diagnoses. A loss with no ``grad_fn`` m

In [10]:
criterion = nn.CrossEntropyLoss(ignore_index=vocab.pad_idx)


def teacher_forced_loss(model, batch):
    """One forward pass, so there is a loss to run backward from.

    `gradient_report` takes an already-computed loss rather than owning the
    forward pass, which keeps it independent of the teacher-forcing convention
    and usable on any model, not just this one.
    """
    source, target = batch
    logits = model(source, target[:, :-1])
    return criterion(logits.reshape(-1, logits.size(-1)), target[:, 1:].reshape(-1))


batch = next(iter(loader(train_df)))

fresh_model = new_model()
print(check_gradients(fresh_model, teacher_forced_loss(fresh_model, batch)))

frozen_encoder = new_model()
for parameter in frozen_encoder.transformer.encoder.parameters():
    parameter.requires_grad_(False)  # <-- the bug
print(check_gradients(frozen_encoder, teacher_forced_loss(frozen_encoder, batch)))

# The pass/fail line is the summary. The report itself names every casualty:
report = gradient_report(frozen_encoder, teacher_forced_loss(frozen_encoder, batch))
print(f"\nfrozen parameters ({len(report.frozen)} of them), first three:")
for name in report.frozen[:3]:
    print(f"  {name}")

[PASS] gradients reach every parameter — live=68 frozen=0 dead=0
[FAIL] gradients reach every parameter — live=42 frozen=26 dead=0; first frozen: transformer.encoder.layers.0.self_attn.in_proj_weight

frozen parameters (26 of them), first three:
  transformer.encoder.layers.0.self_attn.in_proj_weight
  transformer.encoder.layers.0.self_attn.in_proj_bias
  transformer.encoder.layers.0.self_attn.out_proj.weight


**Recognize it:** the loss is flat, and near `ln(V)`.

**Identify it** with the two checks above, in this order:

1. `check_gradients` reports **frozen** parameters → someone called
   `requires_grad_(False)`, or built the optimizer over a filtered parameter
   list.
2. It reports **dead** parameters → the graph is detached somewhere, or that
   output never reaches the loss.
3. Everything is **live** but the loss still doesn't move → gradients exist and
   are being discarded. Check that `optimizer.step()` is actually called, and
   that the learning rate is not zero — which is the bug in the cell above, and
   why it reports every parameter live.

A frozen encoder does *not* flatten the loss; it gives a merely disappointing
model, which is why the gradient check earns its place
([When it fails](../concepts/when-it-fails.md) has the numbers).

## Question 3 — Is it learning the *wrong* thing?

Now the loss falls, beautifully, all the way down. On the training set.

The tell is the *sign* of the gap between training and validation loss. Look at
the healthy run: validation came out **below** training loss. That is not a
fluke — dropout is active during training and disabled during validation, so a
healthy model usually scores slightly better on validation.

When that ordering flips and keeps going, the model is memorizing.

In [11]:
seen = train_df.head(60)  # <-- the bug: far too little data
memorizer = new_model()
memorizer_run = train_model(
    memorizer,
    loader(seen, batch_size=16),
    loader(val_df, shuffle=False),
    num_epochs=150,
    config=CFG,
    optimizer=optim.Adam(memorizer.parameters(), lr=1e-3),
)
print()
print(check_generalization(healthy_run.train_losses, healthy_run.val_losses))
print(check_generalization(memorizer_run.train_losses, memorizer_run.val_losses))

# The single number above is the end of a story. Watch it develop, by passing
# the histories sliced to each epoch:
print("\n  epoch    train      val      gap")
for epoch in (10, 30, 60, 100, 150):
    i = epoch - 1
    train_i, val_i = memorizer_run.train_losses[i], memorizer_run.val_losses[i]
    print(f"  {epoch:5d}    {train_i:.3f}    {val_i:.3f}   {val_i - train_i:+.3f}")

Epoch 1/150 | Train: 5.3491 | Val: 5.3691


Epoch 2/150 | Train: 5.3434 | Val: 5.3612


Epoch 3/150 | Train: 5.3331 | Val: 5.3478


Epoch 4/150 | Train: 5.3531 | Val: 5.3288


Epoch 5/150 | Train: 5.2746 | Val: 5.3042


Epoch 6/150 | Train: 5.2193 | Val: 5.2740


Epoch 7/150 | Train: 5.2525 | Val: 5.2385


Epoch 8/150 | Train: 5.2438 | Val: 5.1980


Epoch 9/150 | Train: 5.1732 | Val: 5.1523


Epoch 10/150 | Train: 5.0675 | Val: 5.1016


Epoch 11/150 | Train: 5.0761 | Val: 5.0466


Epoch 12/150 | Train: 5.0796 | Val: 4.9881


Epoch 13/150 | Train: 4.9723 | Val: 4.9261


Epoch 14/150 | Train: 4.9820 | Val: 4.8609


Epoch 15/150 | Train: 4.8289 | Val: 4.7932


Epoch 16/150 | Train: 4.7687 | Val: 4.7240


Epoch 17/150 | Train: 4.7760 | Val: 4.6535


Epoch 18/150 | Train: 4.7163 | Val: 4.5822


Epoch 19/150 | Train: 4.5989 | Val: 4.5111


Epoch 20/150 | Train: 4.5597 | Val: 4.4416


Epoch 21/150 | Train: 4.4787 | Val: 4.3742


Epoch 22/150 | Train: 4.3796 | Val: 4.3099


Epoch 23/150 | Train: 4.3205 | Val: 4.2483


Epoch 24/150 | Train: 4.2751 | Val: 4.1903


Epoch 25/150 | Train: 4.2020 | Val: 4.1359


Epoch 26/150 | Train: 4.1376 | Val: 4.0838


Epoch 27/150 | Train: 4.1200 | Val: 4.0354


Epoch 28/150 | Train: 4.0479 | Val: 3.9886


Epoch 29/150 | Train: 4.0070 | Val: 3.9435


Epoch 30/150 | Train: 3.9459 | Val: 3.8977


Epoch 31/150 | Train: 3.9032 | Val: 3.8551


Epoch 32/150 | Train: 3.8795 | Val: 3.8161


Epoch 33/150 | Train: 3.8243 | Val: 3.7780


Epoch 34/150 | Train: 3.7965 | Val: 3.7394


Epoch 35/150 | Train: 3.7536 | Val: 3.7041


Epoch 36/150 | Train: 3.7082 | Val: 3.6719


Epoch 37/150 | Train: 3.6823 | Val: 3.6408


Epoch 38/150 | Train: 3.6150 | Val: 3.6098


Epoch 39/150 | Train: 3.6299 | Val: 3.5793


Epoch 40/150 | Train: 3.5897 | Val: 3.5507


Epoch 41/150 | Train: 3.5959 | Val: 3.5222


Epoch 42/150 | Train: 3.5447 | Val: 3.4948


Epoch 43/150 | Train: 3.4429 | Val: 3.4680


Epoch 44/150 | Train: 3.3699 | Val: 3.4421


Epoch 45/150 | Train: 3.3954 | Val: 3.4153


Epoch 46/150 | Train: 3.3605 | Val: 3.3897


Epoch 47/150 | Train: 3.3280 | Val: 3.3647


Epoch 48/150 | Train: 3.2848 | Val: 3.3393


Epoch 49/150 | Train: 3.2749 | Val: 3.3138


Epoch 50/150 | Train: 3.2730 | Val: 3.2885


Epoch 51/150 | Train: 3.2345 | Val: 3.2631


Epoch 52/150 | Train: 3.1534 | Val: 3.2369


Epoch 53/150 | Train: 3.1152 | Val: 3.2111


Epoch 54/150 | Train: 3.1041 | Val: 3.1842


Epoch 55/150 | Train: 3.0599 | Val: 3.1580


Epoch 56/150 | Train: 3.0148 | Val: 3.1323


Epoch 57/150 | Train: 2.9789 | Val: 3.1068


Epoch 58/150 | Train: 2.9842 | Val: 3.0804


Epoch 59/150 | Train: 2.9028 | Val: 3.0520


Epoch 60/150 | Train: 2.9036 | Val: 3.0247


Epoch 61/150 | Train: 2.8493 | Val: 2.9978


Epoch 62/150 | Train: 2.8379 | Val: 2.9726


Epoch 63/150 | Train: 2.8085 | Val: 2.9467


Epoch 64/150 | Train: 2.7490 | Val: 2.9225


Epoch 65/150 | Train: 2.7079 | Val: 2.8992


Epoch 66/150 | Train: 2.6963 | Val: 2.8757


Epoch 67/150 | Train: 2.6090 | Val: 2.8547


Epoch 68/150 | Train: 2.5959 | Val: 2.8329


Epoch 69/150 | Train: 2.6085 | Val: 2.8091


Epoch 70/150 | Train: 2.4941 | Val: 2.7849


Epoch 71/150 | Train: 2.5191 | Val: 2.7598


Epoch 72/150 | Train: 2.4404 | Val: 2.7370


Epoch 73/150 | Train: 2.4305 | Val: 2.7133


Epoch 74/150 | Train: 2.3475 | Val: 2.6903


Epoch 75/150 | Train: 2.3145 | Val: 2.6670


Epoch 76/150 | Train: 2.2985 | Val: 2.6468


Epoch 77/150 | Train: 2.2753 | Val: 2.6274


Epoch 78/150 | Train: 2.2322 | Val: 2.6068


Epoch 79/150 | Train: 2.2316 | Val: 2.5892


Epoch 80/150 | Train: 2.1930 | Val: 2.5675


Epoch 81/150 | Train: 2.0896 | Val: 2.5437


Epoch 82/150 | Train: 2.1101 | Val: 2.5219


Epoch 83/150 | Train: 2.0703 | Val: 2.5022


Epoch 84/150 | Train: 2.0782 | Val: 2.4836


Epoch 85/150 | Train: 2.0231 | Val: 2.4651


Epoch 86/150 | Train: 1.9734 | Val: 2.4458


Epoch 87/150 | Train: 1.9661 | Val: 2.4302


Epoch 88/150 | Train: 1.9168 | Val: 2.4137


Epoch 89/150 | Train: 1.9016 | Val: 2.3981


Epoch 90/150 | Train: 1.8463 | Val: 2.3811


Epoch 91/150 | Train: 1.8651 | Val: 2.3602


Epoch 92/150 | Train: 1.8025 | Val: 2.3422


Epoch 93/150 | Train: 1.7803 | Val: 2.3243


Epoch 94/150 | Train: 1.7409 | Val: 2.3049


Epoch 95/150 | Train: 1.6914 | Val: 2.2834


Epoch 96/150 | Train: 1.7287 | Val: 2.2644


Epoch 97/150 | Train: 1.6673 | Val: 2.2512


Epoch 98/150 | Train: 1.6096 | Val: 2.2403


Epoch 99/150 | Train: 1.5798 | Val: 2.2303


Epoch 100/150 | Train: 1.5537 | Val: 2.2142


Epoch 101/150 | Train: 1.5555 | Val: 2.1948


Epoch 102/150 | Train: 1.4872 | Val: 2.1760


Epoch 103/150 | Train: 1.4656 | Val: 2.1575


Epoch 104/150 | Train: 1.4814 | Val: 2.1392


Epoch 105/150 | Train: 1.4177 | Val: 2.1205


Epoch 106/150 | Train: 1.4117 | Val: 2.1001


Epoch 107/150 | Train: 1.3901 | Val: 2.0824


Epoch 108/150 | Train: 1.3384 | Val: 2.0708


Epoch 109/150 | Train: 1.3250 | Val: 2.0614


Epoch 110/150 | Train: 1.3110 | Val: 2.0442


Epoch 111/150 | Train: 1.2938 | Val: 2.0295


Epoch 112/150 | Train: 1.2361 | Val: 2.0135


Epoch 113/150 | Train: 1.2332 | Val: 1.9989


Epoch 114/150 | Train: 1.2216 | Val: 1.9818


Epoch 115/150 | Train: 1.1713 | Val: 1.9659


Epoch 116/150 | Train: 1.1439 | Val: 1.9505


Epoch 117/150 | Train: 1.1273 | Val: 1.9358


Epoch 118/150 | Train: 1.0822 | Val: 1.9179


Epoch 119/150 | Train: 1.0562 | Val: 1.8991


Epoch 120/150 | Train: 1.0550 | Val: 1.8853


Epoch 121/150 | Train: 1.0366 | Val: 1.8774


Epoch 122/150 | Train: 1.0064 | Val: 1.8661


Epoch 123/150 | Train: 1.0041 | Val: 1.8593


Epoch 124/150 | Train: 0.9792 | Val: 1.8481


Epoch 125/150 | Train: 0.9400 | Val: 1.8339


Epoch 126/150 | Train: 0.9659 | Val: 1.8214


Epoch 127/150 | Train: 0.8740 | Val: 1.8124


Epoch 128/150 | Train: 0.8990 | Val: 1.8034


Epoch 129/150 | Train: 0.8515 | Val: 1.7909


Epoch 130/150 | Train: 0.8227 | Val: 1.7844


Epoch 131/150 | Train: 0.8440 | Val: 1.7772


Epoch 132/150 | Train: 0.7886 | Val: 1.7651


Epoch 133/150 | Train: 0.8248 | Val: 1.7610


Epoch 134/150 | Train: 0.7729 | Val: 1.7586


Epoch 135/150 | Train: 0.7614 | Val: 1.7522


Epoch 136/150 | Train: 0.7668 | Val: 1.7388


Epoch 137/150 | Train: 0.7148 | Val: 1.7254


Epoch 138/150 | Train: 0.6992 | Val: 1.7159


Epoch 139/150 | Train: 0.6886 | Val: 1.7064


Epoch 140/150 | Train: 0.6832 | Val: 1.6979


Epoch 141/150 | Train: 0.6607 | Val: 1.6897


Epoch 142/150 | Train: 0.6217 | Val: 1.6745


Epoch 143/150 | Train: 0.6544 | Val: 1.6668


Epoch 144/150 | Train: 0.5974 | Val: 1.6660


Epoch 145/150 | Train: 0.5880 | Val: 1.6603


Epoch 146/150 | Train: 0.5846 | Val: 1.6511


Epoch 147/150 | Train: 0.5844 | Val: 1.6406


Epoch 148/150 | Train: 0.5428 | Val: 1.6321


Epoch 149/150 | Train: 0.5525 | Val: 1.6265


Epoch 150/150 | Train: 0.5402 | Val: 1.6240

[PASS] generalizes — train=0.265 val=0.099 gap=-0.167 (need <=+0.3)
[FAIL] generalizes — train=0.540 val=1.624 gap=+1.084 (need <=+0.3)

  epoch    train      val      gap
     10    5.068    5.102   +0.034
     30    3.946    3.898   -0.048
     60    2.904    3.025   +0.121
    100    1.554    2.214   +0.660
    150    0.540    1.624   +1.084


**Recognize it:** validation loss sits clearly *above* training loss, and the gap
widens as training continues — in the table above it runs +0.03, +0.12, +0.66,
+1.08. Watching only the training curve hides this completely: that curve looks
better than the healthy model's, all the way down.

At epochs 10 and 30 the gap is near zero, even negative, and this run looks like
the healthy one. Overfitting is something a model does over time, which is why
you watch the gap rather than test it once.

**Fix it:** more data first; then regularization (dropout, label smoothing) and
early stopping on validation loss. Note that early stopping requires a
validation set the model never trains on — which is the subject of the next
question.

## Question 4 — Is the measurement lying?

The most dangerous failure is the one that makes your numbers look *good*.

The model from Question 3 memorized 60 sentence pairs. Score it on those same
pairs and it looks outstanding. Score it on fresh ones and it collapses. Both
numbers come from the same model and the same metric.

In [12]:
def bleu_on(model, frame):
    return compute_bleu(translate(model, list(frame.src)), list(frame.tgt)).score


fresh = make_corpus(60, seed=4321)
mixed = pd.concat([seen, fresh]).reset_index(drop=True)

print(f"BLEU on the 60 pairs it trained on : {bleu_on(memorizer, seen):5.2f}")
print(f"BLEU on 60 pairs it has never seen : {bleu_on(memorizer, fresh):5.2f}")
print(f"BLEU on a 50/50 mix of the two     : {bleu_on(memorizer, mixed):5.2f}   <- true of neither")

BLEU on the 60 pairs it trained on : 81.81
BLEU on 60 pairs it has never seen :  6.53


BLEU on a 50/50 mix of the two     : 47.83   <- true of neither


A contaminated test set does not report an error. It reports a number, halfway
between the truth and a lie, and you have no way to tell from the number alone.

**The diagnostic** is to compare the sets directly, before trusting any score:

In [13]:
# check_contamination takes the source-side sentences of each split, not frames,
# so it works whatever your corpus is held in.
print(check_contamination(mixed.src, seen.src))
print(check_contamination(fresh.src, seen.src))

[FAIL] test set is clean — 60/120 test sources also appear in training; e.g. 'Ana likes 18 young fish', 'Ana likes 21 birds', 'Ana sees 48 big birds'
[PASS] test set is clean — 0/60 test sources also appear in training


**Recognize it:** a score that is much better than the model's behaviour when you
read its actual output, or a suspicious jump after a data change.

**Fix it:** split before you do anything else, deduplicate across the split, and
re-run the overlap check whenever the corpus changes. Deduplicate on the source
side at minimum — near-duplicates that differ only in whitespace or casing will
slip past the exact-match check above.

!!! warning "This generalizes beyond contamination"
    Any comparison where more than one thing changed produces a number that is
    true of nothing. Change one variable at a time, or you are measuring their sum.

## Question 5 — Is it the environment?

Everything above is about the model. This one is about the seven lines around it.

The canonical version: inference with the model still in training mode. Dropout
stays active, so it randomly zeroes activations and rescales the survivors. The
model still runs. It still produces plausible output. Every number it gives you
is wrong, and a different kind of wrong each time you ask.

In [14]:
def validation_loss(model):
    total = batches = 0
    for source, target in loader(val_df, shuffle=False):
        with torch.no_grad():
            logits = model(source, target[:, :-1])
        total += criterion(
            logits.reshape(-1, logits.size(-1)), target[:, 1:].reshape(-1)
        ).item()
        batches += 1
    return total / batches


healthy.train()
in_train_mode = [round(validation_loss(healthy), 4) for _ in range(3)]
healthy.eval()
in_eval_mode = [round(validation_loss(healthy), 4) for _ in range(3)]

print(f"same model, same data, train() mode: {in_train_mode}")
print(f"same model, same data, eval()  mode: {in_eval_mode}")

print()
healthy.train()
print(check_eval_mode(healthy))
healthy.eval()
print(check_eval_mode(healthy))

# A CheckResult is truthy when it passed, so it reads as a guard too:
assert check_eval_mode(healthy), "refusing to score a model in training mode"

same model, same data, train() mode: [0.2639, 0.2547, 0.2534]
same model, same data, eval()  mode: [0.0987, 0.0987, 0.0987]

[FAIL] model is in eval mode — model.training=True, so dropout is active and results will not reproduce
[PASS] model is in eval mode — model.training=False, so results are deterministic


**Recognize it:** two identical runs disagree. That is the signature, and it is
worth more than any single number — a measurement you cannot repeat is not a
measurement. Note that the train-mode loss here is not merely noisy, it is
roughly **2.5x** the true value, so it is also badly biased.

**Fix it:** call `model.eval()` before any evaluation, decoding, or attention
plot, and wrap inference in `torch.no_grad()`. If you need a stronger guarantee,
assert it — `check_eval_mode` above is two lines.

The same class of bug covers a stale checkpoint, a model on a different device
than its inputs, and a vocabulary rebuilt after the checkpoint was saved. Each
one leaves the code running and the results meaningless.

## The procedure

| # | Question | Symptom | Check | Fix |
| - | -------- | ------- | ----- | --- |
| 1 | Is the data right? | loss falls but plateaus high; output unrelated to input | `diagnose_alignment(frame).looks_aligned()` | realign or rebuild the corpus |
| 2 | Is it learning at all? | loss flat near `ln(V)` | `check_loss_moved`, then `check_gradients` | unfreeze / reattach / fix the optimizer |
| 3 | Is it learning the wrong thing? | val loss above train loss, gap widening | `check_generalization` | more data, regularization, early stopping |
| 4 | Is the measurement lying? | scores too good for the output you read | `check_contamination` | split and deduplicate first |
| 5 | Is it the environment? | two identical runs disagree | `check_eval_mode` | `model.eval()` and `torch.no_grad()` |

All but the first live in [`torchlingo.diagnostics`](../reference/diagnostics.md).
The first needs no model at all, so it sits with the corpus tools in
`torchlingo.preprocessing.alignment` — which is also why it is the cheapest
question to ask.

Three habits are worth more than the table:

1. **Know your reference points.** `ln(V)` is what guessing scores. A healthy
   validation loss sits at or slightly below training loss. Without those you
   cannot tell a bad number from a fine one.
2. **Keep a control.** Nearly every check above is a comparison. The single most
   useful artifact when something breaks is a run you know was healthy.
3. **Change one thing at a time.** Two changes and one number measures their sum,
   which is true of neither.

### Try it yourself

The checks are the point, so use them on a break you haven't seen:

- Swap `src` and `tgt` for half the corpus. Which check catches it? (Note that
  length correlation will not — both sides stay the same length.)
- Build the vocabulary from `train_df` only, then evaluate on `val_df`, whose
  numbers include types the vocabulary has never seen.
- Detach the encoder output before it reaches the decoder, and see which bucket
  `gradient_report` puts the encoder's parameters in. Is it the same one
  freezing them produced?

For the failures behind these checks, read
[When it fails](../concepts/when-it-fails.md).